File: app/src/main/java/cm/aptoide/pt/search/view/item/SearchResultViewHolder.java
package cm.aptoide.pt.search.view.item;

import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import cm.aptoide.aptoideviews.downloadprogressview.DownloadProgressView;
import cm.aptoide.pt.R;
import cm.aptoide.pt.download.view.DownloadClick;
import cm.aptoide.pt.download.view.DownloadEvent;
import cm.aptoide.pt.download.view.DownloadStatusModel;
import cm.aptoide.pt.download.view.DownloadViewStatusHelper;
import cm.aptoide.pt.home.AppSecondaryInfoViewHolder;
import cm.aptoide.pt.networking.image.ImageLoader;
import cm.aptoide.pt.search.model.SearchAppResult;
import cm.aptoide.pt.search.model.SearchAppResultWrapper;
import cm.aptoide.pt.utils.AptoideUtils;
import java.text.DecimalFormat;
import rx.subjects.PublishSubject;

public class SearchResultViewHolder extends SearchResultItemView<SearchAppResult> {

  public static final int LAYOUT = R.layout.search_app_row;

  private final PublishSubject<SearchAppResultWrapper> onItemViewClick;
  private final PublishSubject<DownloadClick> downloadClickPublishSubject;
  private final AppSecondaryInfoViewHolder appInfoViewHolder;

  private final String query;
  private TextView nameTextView;
  private ImageView iconImageView;
  private TextView downloadsTextView;
  private TextView ratingBar;
  private TextView storeTextView;
  private SearchAppResult searchApp;
  private Button installButton;
  private DownloadProgressView downloadProgressView;

  private DownloadViewStatusHelper downloadViewStatusHelper;

  public SearchResultViewHolder(View itemView,
      PublishSubject<SearchAppResultWrapper> onItemViewClick,
      PublishSubject<DownloadClick> downloadClickPublishSubject, String query) {
    super(itemView);
    this.onItemViewClick = onItemViewClick;
    this.query = query;
    this.downloadClickPublishSubject = downloadClickPublishSubject;
    this.downloadViewStatusHelper = new DownloadViewStatusHelper(itemView.getContext());
    appInfoViewHolder = new AppSecondaryInfoViewHolder(itemView, new DecimalFormat("0.0"));
    bindViews(itemView);
  }

  @Override public void setup(SearchAppResult result) {
    this.searchApp = result;
    appInfoViewHolder.setInfo(result.hasBilling() || result.hasAdvertising(),
        result.getAverageRating(), false, false);
    setAppName();
    setDownloadCount();
    setAverageValue();
    setStoreName();
    setIconView();
    setDownloadStatus(searchApp);
  }

  public void setDownloadStatus(SearchAppResult app) {
    DownloadStatusModel downloadModel = app.getDownloadModel();
    if (app.isHighlightedResult() && downloadModel != null) {
      downloadViewStatusHelper.setDownloadStatus(app, installButton, downloadProgressView);
    } else {
      installButton.setVisibility(View.GONE);
      downloadProgressView.setVisibility(View.GONE);
    }
  }

  private void setIconView() {
    ImageLoader.with(iconImageView.getContext())
        .load(searchApp.getIcon(), iconImageView);
  }

  private void setStoreName() {
    storeTextView.setText(searchApp.getStoreName());
  }

  private void setAverageValue() {
    float avg = searchApp.getAverageRating();
    if (avg <= 0) {
      ratingBar.setText(R.string.appcardview_title_no_stars);
    } else {
      ratingBar.setVisibility(View.VISIBLE);
      ratingBar.setText(new DecimalFormat("0.0").format(avg));
    }
  }

  private void setDownloadCount() {
    downloadsTextView.setText(AptoideUtils.StringU.withSuffix(searchApp.getTotalDownloads()));
  }

  private void setAppName() {
    nameTextView.setText(searchApp.getAppName());
  }

  private void bindViews(View itemView) {
    nameTextView = itemView.findViewById(R.id.app_name);
    iconImageView = itemView.findViewById(R.id.app_icon);
    downloadsTextView = itemView.findViewById(R.id.downloads);
    ratingBar = itemView.findViewById(R.id.rating);
    storeTextView = itemView.findViewById(R.id.store_name);
    installButton = itemView.findViewById(R.id.install_button);
    downloadProgressView = itemView.findViewById(R.id.download_progress_view);

    itemView.setOnClickListener(v -> {
      onItemViewClick.onNext(new SearchAppResultWrapper(query, searchApp, getAdapterPosition()));
    });
    installButton.setOnClickListener(v -> {
      downloadClickPublishSubject.onNext(new DownloadClick(searchApp, DownloadEvent.INSTALL));
    });
    downloadProgressView.setEventListener(action -> {
      switch (action.getType()) {
        case CANCEL:
          downloadClickPublishSubject.onNext(new DownloadClick(searchApp, DownloadEvent.CANCEL));
          break;
        case RESUME:
          downloadClickPublishSubject.onNext(new DownloadClick(searchApp, DownloadEvent.RESUME));
          break;
        case PAUSE:
          downloadClickPublishSubject.onNext(new DownloadClick(searchApp, DownloadEvent.PAUSE));
          break;
      }
    });
  }
}

File: app/src/main/java/cm/aptoide/pt/search/view/item/SearchResultViewHolder.kt
package cm.aptoide.pt.search.view.item

import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import cm.aptoide.pt.R
import cm.aptoide.pt.app.view.screenshots.ScreenShotClickEvent
import cm.aptoide.pt.app.view.screenshots.ScreenshotsAdapter
import cm.aptoide.pt.dataprovider.model.v7.Malware
import cm.aptoide.pt.download.view.DownloadClick
import cm.aptoide.pt.download.view.DownloadViewStatusHelper
import cm.aptoide.pt.home.AppSecondaryInfoViewHolder
import cm.aptoide.pt.networking.image.ImageLoader
import cm.aptoide.pt.search.model.SearchAppResult
import cm.aptoide.pt.search.model.SearchAppResultWrapper
import cm.aptoide.pt.utils.AptoideUtils
import cm.aptoide.pt.view.app.AppScreenshot
import kotlinx.android.synthetic.main.search_app_row.view.*
import rx.subjects.PublishSubject
import java.text.DecimalFormat
import java.util.*

class SearchResultViewHolder(itemView: View,
                             private val itemClickSubject: PublishSubject<SearchAppResultWrapper>,
                             private val downloadClickSubject: PublishSubject<DownloadClick>,
                             private val screenShotClick: PublishSubject<ScreenShotClickEvent>,
                             private val query: String?) :
    SearchResultItemView<SearchAppResult?>(itemView) {

  private val downloadViewStatusHelper = DownloadViewStatusHelper(itemView.context)

  private val appInfoViewHolder: AppSecondaryInfoViewHolder =
      AppSecondaryInfoViewHolder(itemView, DecimalFormat("0.0"))

  private val adapter =
      ScreenshotsAdapter(Collections.emptyList(), Collections.emptyList(), screenShotClick, 128)

  init {
    itemView.media_rv.layoutManager =
        LinearLayoutManager(itemView.context, RecyclerView.HORIZONTAL, false)
    itemView.media_rv.isNestedScrollingEnabled = false
    itemView.media_rv.adapter = adapter
  }

  override fun setup(result: SearchAppResult?) {
    result?.let {
      setAppInfo(result)
      setDownloadStatus(result)
    }
  }

  fun setDownloadStatus(app: SearchAppResult) {
    val downloadModel = app.downloadModel
    if (app.isHighlightedResult && downloadModel != null) {
      downloadViewStatusHelper.setDownloadStatus(app.download, itemView.install_button,
          itemView.download_progress_view)
      setupMediaAdapter(app.screenshots)
      itemView.media_rv.visibility = View.VISIBLE
    } else {
      itemView.install_button.visibility = View.GONE
      itemView.download_progress_view.visibility = View.GONE
      itemView.media_rv.visibility = View.GONE
    }
  }

  private fun setupMediaAdapter(screenshots: List<AppScreenshot>) {
    adapter.updateScreenshots(screenshots)
    adapter.updateVideos(Collections.emptyList())
  }

  private fun setAppInfo(result: SearchAppResult) {
    itemView.app_name.text = result.appName
    itemView.downloads.text = result.totalDownloads.let { AptoideUtils.StringU.withSuffix(it) }
    ImageLoader.with(itemView.app_icon.context).load(result.icon, itemView.app_icon)

    val avgRating = result.averageRating
    if (avgRating <= 0) {
      itemView.rating.setText(R.string.appcardview_title_no_stars)
    } else {
      itemView.rating.visibility = View.VISIBLE
      itemView.rating.text = DecimalFormat("0.0").format(avgRating.toDouble())
    }

    itemView.store_name.text = result.storeName
    appInfoViewHolder.setInfo(result.hasBilling() || result.hasAdvertising(), result.averageRating,
        false, false)

    when (result.rank) {
      Malware.Rank.TRUSTED.ordinal -> {
        ImageLoader.with(itemView.app_badge.context)
            .load(R.drawable.ic_badges_trusted, itemView.app_badge)
      }
      Malware.Rank.CRITICAL.ordinal -> {
        ImageLoader.with(itemView.app_badge.context)
            .load(R.drawable.ic_badges_critical, itemView.app_badge)
      }
      Malware.Rank.WARNING.ordinal -> {
        ImageLoader.with(itemView.app_badge.context)
            .load(R.drawable.ic_badges_warning, itemView.app_badge)
      }
      Malware.Rank.UNKNOWN.ordinal -> {
        ImageLoader.with(itemView.app_badge.context)
            .load(R.drawable.ic_badges_unknown, itemView.app_badge)
      }
    }

    itemView.setOnClickListener {
      itemClickSubject.onNext(SearchAppResultWrapper(query, result, adapterPosition))
    }
    downloadViewStatusHelper.setupListeners(result.download, downloadClickSubject,
        itemView.install_button,
        itemView.download_progress_view)
  }

  companion object {
    const val LAYOUT = R.layout.search_app_row
  }
}